homeassistant/components/onvif/camera.py

```diff
@@ -304,40 +354,76 @@ def __init__(self, hass, config):
         except exceptions.ONVIFError as err:
             _LOGGER.error("Couldn't setup camera '%s'. Error: %s", self._name, err)
 
-    def setup_ptz(self):
+    async def setup_ptz(self):
```

Comment:
Why change here to a coroutine?

Reply:
Indeed, did that too quickly after reading a post in issue #27744 saying it solved an issue.
I will remove it if unecessary.

Thank you for all your comments, actually i'm rewriting a new onvif2 component which will have more complete support on other PTZ operations, i will address your change request in it.
I was thinking first to release it as a custom_component before to make a new PR when it will be widely tested and mature. This way I would not consume your time day after day on code review.
If you, Team guys, also think it would be a better way, i might better close that PR then.

Reply:
Further changes to this integration are welcome but not in this PR. Please make a new separate PR for further changes.

Reply:
@olijouve 

> actually i'm rewriting a new onvif2 component which will have more complete support on other PTZ operations, i will address your change request in it.

One change which might make onvif2 version more flexible would be to define PTZ parameters as floats, rather than the hard coded UP/DOWN/LEFT/RIGHT/IN/OUT.  And possibly second to retrieve parameter min/max ranges via camera ONVIF request.

> release it as a custom_component 

I am not a HA coder (at least not yet :-).  And I have been copying your GIT code to custom_component to test, all along.  I don't know yet how to do it otherwise.  So this method works fine for me.

Reply:
> Why change here to a coroutine?

@MartinHjelmare almost all others methods of the OnvifCamera class are async coroutines they should not be ?

Reply:
We only want to define coroutines when we need to await something inside the function. There's an overhead time wise to await a coroutine versus calling a normal function. So we try to be conservative.

Reply:
@MartinHjelmare 

I am a C & C++ embedded developer, but not on HA or Python developer.

Just curious... it seems a local function is synchronous and blocking... a coroutine is asynchronous and non-blocking?  When a function requires command & response to a remote device like a camera, wouldn't that be valid for coroutine (non-blocking) function.  Or is the integration (camera/onvif) already executing in a coroutine?

Reply:
It depends on the device library if it supports asyncio or not and what api it exposes.